framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -890,6 +931,23 @@ public Void visitTypeParameter(TypeParameterTree node, Void p) {
 
     @Override
     public Void visitVariable(VariableTree node, Void p) {
+
+        // Warn about type annotations written before modifiers such as "public".  javac retains no
+        // information about modifier locations.  So, this is a very partial check:  Issue a warning
+        // if a type annotation is at the very beginning of the VariableTree, and a modifer follows
+        // it.
+        ModifiersTree modifiersTree = node.getModifiers();
+        Set<Modifier> modifierSet = modifiersTree.getFlags();
+        List<? extends AnnotationTree> annotations = modifiersTree.getAnnotations();
+        if (!modifierSet.isEmpty() && !annotations.isEmpty()) {
+            AnnotationTree anno = annotations.get(0);
+            if (((JCTree) anno).getStartPosition() == ((JCTree) node).getStartPosition()
```

Comment:
Would something like this be a more general (yet still hacky) check:
- convert modifierSet into a String, so we get "public static final" or some such.
- keep a counter `nextPos` initialized to that string's length.
- go through each annotation:
  - if it's a declaration annotation: add the length of the annotation to nextPos
    - length of toString of the annotation isn't useful, as it might be different length than original input
    - maybe using JCTree.getEndPosition(EndPosTable) and com.sun.tools.javac.tree.EndPosTable, note requires -Xjcov
  - if it's a type annotation: make sure start pos of annotation > nextPos.
  - once you see a type annotation, don't allow declaration annotations.

Ensuring that no declaration annotation happens after a type annotation should be an easy additional check.
As precise calculations of lengths would require -Xjcov, I'm not sure we would want to do that in the BaseTypeVisitor, but rather write a separate optional checker for it.